app/src/main/java/io/gnosis/safe/ui/base/BaseStateViewModel.kt

```diff
@@ -16,7 +20,36 @@ data class AppDispatchers(
     val background: CoroutineDispatcher = Dispatchers.IO
 )
 
-abstract class BaseStateViewModel<T>(private val dispatchers: AppDispatchers) : ViewModel() where T : BaseStateViewModel.State {
+abstract class BaseStateViewModel<T>(private val dispatchers: AppDispatchers, private val connectivityManager: ConnectivityManager? = null) :
+    ViewModel() where T : BaseStateViewModel.State {
+
+    private lateinit var networkCallback: ConnectivityManager.NetworkCallback
+
+    init {
+        connectivityManager?.let {
+            networkCallback = object : ConnectivityManager.NetworkCallback() {
+                override fun onAvailable(network: Network?) {
+                    super.onAvailable(network)
+                    safeLaunch {
+                        updateState(true) { viewAction = ViewAction.Connectivity(false); this }
+                    }
+                }
+
+                override fun onLost(network: Network?) {
+                    super.onLost(network)
+                    safeLaunch {
+                        updateState(true) { viewAction = ViewAction.Connectivity(true); this }
+                    }
+                }
+            }
+            val builder = NetworkRequest.Builder()
+            builder
+                .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+                .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
+            it.registerNetworkCallback(builder.build(), networkCallback)
+        }
+    }
```

Comment:
Oh man not sure this is the best decision. I would absolutely **not** couple the `ConnectivityManager` to our BaseViewModel class. This doesn't match our modelling at all. It feels put here out of convenience rather than architectural design.

Reply:
could you elaborate more? Why does it not match our modeling? Why having ConnectivityManager provided to a viewModel is a bad thing architecturally?

Reply:
what do you mean by a forceful inheritance? Why would we need to poll the connectivity manager?